pkg/reconciler/pipeline/dag/dag.go

```diff
@@ -121,40 +126,62 @@ func GetCandidateTasks(g *Graph, doneTasks ...string) (sets.String, error) {
 }
 
 func linkPipelineTasks(prev *Node, next *Node) error {
-	// Check for self cycle
-	if prev.Task.HashKey() == next.Task.HashKey() {
-		return fmt.Errorf("cycle detected; task %q depends on itself", next.Task.HashKey())
-	}
-	// Check if we are adding cycles.
-	path := []string{next.Task.HashKey(), prev.Task.HashKey()}
-	if err := lookForNode(prev.Prev, path, next.Task.HashKey()); err != nil {
-		return fmt.Errorf("cycle detected: %w", err)
-	}
 	next.Prev = append(next.Prev, prev)
 	prev.Next = append(prev.Next, next)
 	return nil
 }
 
-func lookForNode(nodes []*Node, path []string, next string) error {
-	for _, n := range nodes {
-		path = append(path, n.Task.HashKey())
-		if n.Task.HashKey() == next {
-			return errors.New(getVisitedPath(path))
+// use Kahn's algorithm to find cycles in dependencies
+func findCyclesInDependencies(deps map[string][]string) error {
+	independentTasks := sets.NewString()
+	dag := make(map[string]sets.String, len(deps))
+	childMap := make(map[string]sets.String, len(deps))
+	for task, taskDeps := range deps {
+		dag[task] = sets.NewString(taskDeps...)
+		for _, dep := range taskDeps {
+			if len(deps[dep]) == 0 {
+				independentTasks.Insert(dep)
+			}
+			if children, ok := childMap[dep]; ok {
+				children.Insert(task)
+			} else {
+				childMap[dep] = sets.NewString(task)
+			}
 		}
-		if err := lookForNode(n.Prev, path, next); err != nil {
-			return err
+	}
+
+	for {
+		parent, ok := independentTasks.PopAny()
+		if !ok {
+			break
+		}
+		children := childMap[parent]
+		for {
+			child, ok := children.PopAny()
+			if !ok {
+				break
+			}
+			dag[child].Delete(parent)
+			if dag[child].Len() == 0 {
+				independentTasks.Insert(child)
+				delete(dag, child)
+			}
 		}
 	}
-	return nil
+
+	return getInterdependencyError(dag)
 }
 
-func getVisitedPath(path []string) string {
-	// Reverse the path since we traversed the Graph using prev pointers.
-	for i := len(path)/2 - 1; i >= 0; i-- {
-		opp := len(path) - 1 - i
-		path[i], path[opp] = path[opp], path[i]
+func getInterdependencyError(dag map[string]sets.String) error {
+	for parent, children := range dag {
+		// report first interdependency
+		var childNames []string
+		for child := range children {
```

Comment:
...which is why `children` here can be empty and therefore the first independent task will be reported in the error.

It should be enough to either remove the independent tasks from the dag (just like children are removed after no entering edges are left), or adding a simple `if` here:
```go
if len(children) == 0 {
  continue
}
```

The Kahn's algorithm is defined for edges, so it doesn't really matter which method will be used as long as it's actually only triggered for existing edges. However, it's easier to define the invariant of `dag` if it only contains tasks that have some edges entering into them. That would require adding:
```go
delete(dag, parent)
```
...right after the `if !ok`. `delete(dag, child)` may be removed then, but it should be safe to leave it be too, as Go supports delete-non-existing for maps.